From 56537410fedad20549d98f95c10b8177fdf8f0ee Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 8 Feb 2023 12:27:40 +0100
Subject: [PATCH 0347/1695] drivers: stm32_exti: add RIF support for EXTI

Add RIF configuration for EXTI, which is a RIF aware device. It
means that the EXTI driver is in charge of configuring its own RIF
restrictions and that the EXTI has dedicated RIF configuration
registers.

RIF configuration is part of private data in struct stm32_exti_priv.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f2974372eee05790935f5bb649ccbcd6468ee0c5 drivers: stm32_exti: add RIF support for EXTI]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I3eea88541200c8e4f78f1edc0955f3d8672c4fe1
---
 core/drivers/stm32_exti.c | 228 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 227 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
index c6b0f3d15..998b64195 100644
--- a/core/drivers/stm32_exti.c
+++ b/core/drivers/stm32_exti.c
@@ -4,6 +4,7 @@
  */
 
 #include <drivers/stm32_exti.h>
+#include <drivers/stm32_rif.h>
 #include <io.h>
 #include <kernel/boot.h>
 #include <kernel/dt.h>
@@ -21,8 +22,31 @@
 #define _EXTI_RPR(n)		(0x00cU + (n) * 0x20U)
 #define _EXTI_FPR(n)		(0x010U + (n) * 0x20U)
 #define _EXTI_SECCFGR(n)	(0x014U + (n) * 0x20U)
+#define _EXTI_PRIVCFGR(n)	(0x018U + (n) * 0x20U)
 #define _EXTI_CR(n)		(0x060U + (n) * 4U)
 #define _EXTI_C1IMR(n)		(0x080U + (n) * 0x10U)
+#define _EXTI_EnCIDCFGR(n)	(0x180U + (n) * 4U)
+#define _EXTI_CmCIDCFGR(n)	(0x300U + (n) * 4U)
+#define _EXTI_HWCFGR1		0x3f0U
+
+/* SECCFGR register bitfields */
+#define _EXTI_SECCFGR_MASK		GENMASK_32(31, 0)
+
+/* PRIVCFGR register bitfields */
+#define _EXTI_PRIVCFGR_MASK		GENMASK_32(31, 0)
+
+/* CIDCFGR register bitfields */
+#define _EXTI_CIDCFGR_SCID_MASK		GENMASK_32(6, 4)
+#define _EXTI_CIDCFGR_CONF_MASK		(_CIDCFGR_CFEN | \
+					 _EXTI_CIDCFGR_SCID_MASK)
+
+/* _EXTI_HWCFGR1 bit fields */
+#define _EXTI_HWCFGR1_NBEVENTS_MASK	GENMASK_32(7, 0)
+#define _EXTI_HWCFGR1_NBEVENTS_SHIFT	0U
+#define _EXTI_HWCFGR1_NBCPUS_MASK	GENMASK_32(11, 8)
+#define _EXTI_HWCFGR1_NBCPUS_SHIFT	8U
+#define _EXTI_HWCFGR1_CIDWIDTH_MASK	GENMASK_32(27, 24)
+#define _EXTI_HWCFGR1_CIDWIDTH_SHIFT	24U
 
 #define _EXTI_MAX_CR		4U
 #define _EXTI_BANK_NR		3U
@@ -30,12 +54,18 @@
 struct stm32_exti_pdata {
 	vaddr_t base;
 	unsigned int lock;
+	uint32_t hwcfgr1;
 	uint32_t wake_active[_EXTI_BANK_NR];
 	uint32_t mask_cache[_EXTI_BANK_NR];
+	uint32_t seccfgr_cache[_EXTI_BANK_NR];
+	uint32_t privcfgr_cache[_EXTI_BANK_NR];
+	uint32_t access_mask[_EXTI_BANK_NR];
 	uint32_t rtsr_cache[_EXTI_BANK_NR];
 	uint32_t ftsr_cache[_EXTI_BANK_NR];
-	uint32_t seccfgr_cache[_EXTI_BANK_NR];
 	uint32_t port_sel_cache[_EXTI_MAX_CR];
+	uint32_t *e_cids;
+	uint32_t *c_cids;
+
 	SLIST_ENTRY(stm32_exti_pdata) link;
 };
 
@@ -58,6 +88,30 @@ static uint32_t stm32_exti_get_bank(uint32_t exti_line)
 	return bank;
 }
 
+static inline uint32_t stm32_exti_maxcid(const struct stm32_exti_pdata *exti)
+{
+	uint32_t bitfield = (exti->hwcfgr1 & _EXTI_HWCFGR1_CIDWIDTH_MASK) >>
+			    _EXTI_HWCFGR1_CIDWIDTH_SHIFT;
+
+	return BIT(bitfield) - 1;
+}
+
+static inline uint32_t stm32_exti_nbevents(const struct stm32_exti_pdata *exti)
+{
+	uint32_t bitfield = (exti->hwcfgr1 & _EXTI_HWCFGR1_NBEVENTS_MASK) >>
+			    _EXTI_HWCFGR1_NBEVENTS_SHIFT;
+
+	return bitfield + 1;
+}
+
+static inline uint32_t stm32_exti_nbcpus(const struct stm32_exti_pdata *exti)
+{
+	uint32_t bitfield = (exti->hwcfgr1 & _EXTI_HWCFGR1_NBCPUS_MASK) >>
+			    _EXTI_HWCFGR1_NBCPUS_SHIFT;
+
+	return bitfield + 1;
+}
+
 void stm32_exti_set_type(struct stm32_exti_pdata *exti, uint32_t exti_line,
 			 uint32_t type)
 {
@@ -209,12 +263,171 @@ void stm32_exti_set_gpio_port_sel(struct stm32_exti_pdata *exti, uint8_t bank,
 	cpu_spin_unlock_xrestore(&exti->lock, exceptions);
 }
 
+static void stm32_exti_rif_parse_dt(struct stm32_exti_pdata *exti,
+				    const void *fdt, int node)
+{
+	struct rif_conf_data conf_data = { };
+	const fdt32_t *cuint = NULL;
+	uint32_t rif_conf = 0;
+	unsigned int i = 0;
+	int len = 0;
+
+	cuint = fdt_getprop(fdt, node, "st,protreg", &len);
+	if (!cuint)
+		panic("No RIF configuration available");
+
+	exti->e_cids = calloc(stm32_exti_nbevents(exti), sizeof(uint32_t));
+	exti->c_cids = calloc(stm32_exti_nbcpus(exti), sizeof(uint32_t));
+	if (!exti->e_cids || !exti->c_cids)
+		panic("Out of memory");
+
+	conf_data.cid_confs   = exti->e_cids;
+	conf_data.sec_conf    = exti->seccfgr_cache;
+	conf_data.priv_conf   = exti->privcfgr_cache;
+	conf_data.access_mask = exti->access_mask;
+
+	for (i = 0; i < len / sizeof(uint32_t); i++) {
+		rif_conf = fdt32_to_cpu(cuint[i]);
+
+		stm32_rif_parse_cfg(rif_conf, &conf_data,
+				    stm32_exti_maxcid(exti),
+				    stm32_exti_nbevents(exti));
+	}
+
+	cuint = fdt_getprop(fdt, node, "st,proccid", &len);
+	if (!cuint)
+		panic("No RIF configuration available");
+
+	for (i = 0; i < len / (2 * sizeof(uint32_t)); i++) {
+		unsigned int pos = fdt32_to_cpu(cuint[2 * i]);
+		uint32_t c_cid = fdt32_to_cpu(cuint[2 * i + 1]);
+
+		if (pos == 0 || pos > stm32_exti_nbcpus(exti))
+			panic("CID position out of range");
+
+		if (c_cid > stm32_exti_maxcid(exti))
+			panic("CID out of range");
+
+		exti->c_cids[pos - 1] = (c_cid << SCID_SHIFT) | _CIDCFGR_CFEN;
+	}
+}
+
+static TEE_Result stm32_exti_rif_apply(const struct stm32_exti_pdata *exti)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int bit_offset = 0;
+	bool is_tdcid = false;
+	uint32_t event = 0;
+	unsigned int i = 0;
+
+	res = stm32_rifsc_check_tdcid(&is_tdcid);
+	if (res)
+		return res;
+
+	/*
+	 * If TDCID, clear EnCIDCFGR and CmCIDCFGR to prevent undesired
+	 * events during the following configuration.
+	 */
+
+	if (is_tdcid) {
+		for (event = 0; event < stm32_exti_nbevents(exti); event++) {
+			i = stm32_exti_get_bank(event);
+			bit_offset = event % 32;
+
+			if (!(BIT(bit_offset) & exti->access_mask[i]))
+				continue;
+
+			io_clrbits32(exti->base + _EXTI_EnCIDCFGR(event),
+				     _EXTI_CIDCFGR_CONF_MASK);
+		}
+
+		for (i = 0; i < stm32_exti_nbcpus(exti); i++)
+			io_clrbits32(exti->base + _EXTI_CmCIDCFGR(i),
+				     _EXTI_CIDCFGR_CONF_MASK);
+	}
+
+	/* Security and privilege RIF configuration */
+	for (i = 0; i < _EXTI_BANK_NR; i++) {
+		if (!exti->access_mask[i])
+			continue;
+
+		io_clrsetbits32(exti->base + _EXTI_PRIVCFGR(i),
+				_EXTI_PRIVCFGR_MASK & exti->access_mask[i],
+				exti->privcfgr_cache[i]);
+		io_clrsetbits32(exti->base + _EXTI_SECCFGR(i),
+				_EXTI_SECCFGR_MASK & exti->access_mask[i],
+				exti->seccfgr_cache[i]);
+	}
+
+	if (!is_tdcid)
+		return TEE_SUCCESS;
+
+	/* If TDCID, configure EnCIDCFGR and CmCIDCFGR */
+	for (event = 0; event < stm32_exti_nbevents(exti); event++) {
+		i = stm32_exti_get_bank(event);
+		bit_offset = event % 32;
+
+		if (!(BIT(bit_offset) & exti->access_mask[i]))
+			continue;
+
+		io_clrsetbits32(exti->base + _EXTI_EnCIDCFGR(event),
+				_EXTI_CIDCFGR_CONF_MASK, exti->e_cids[event]);
+	}
+	for (i = 0; i < stm32_exti_nbcpus(exti); i++) {
+		if (!(exti->c_cids[i] & _CIDCFGR_CFEN))
+			continue;
+
+		io_clrsetbits32(exti->base + _EXTI_CmCIDCFGR(i),
+				_EXTI_CIDCFGR_CONF_MASK, exti->c_cids[i]);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void stm32_exti_rif_save(struct stm32_exti_pdata *exti)
+{
+	unsigned int bit_offset = 0;
+	bool is_tdcid = false;
+	uint32_t event = 0;
+	unsigned int i = 0;
+
+	for (i = 0; i < _EXTI_BANK_NR; i++) {
+		if (!exti->access_mask[i])
+			continue;
+
+		exti->privcfgr_cache[i] =
+			io_read32(exti->base + _EXTI_PRIVCFGR(i));
+		exti->seccfgr_cache[i] =
+			io_read32(exti->base + _EXTI_SECCFGR(i));
+	}
+
+	stm32_rifsc_check_tdcid(&is_tdcid);
+	if (!is_tdcid)
+		return;
+
+	for (event = 0; event < stm32_exti_nbevents(exti); event++) {
+		i = stm32_exti_get_bank(event);
+		bit_offset = event % 32;
+
+		if (!(BIT(bit_offset) & exti->access_mask[i]))
+			continue;
+
+		exti->e_cids[event] = io_read32(exti->base +
+						_EXTI_EnCIDCFGR(event));
+	}
+	for (i = 0; i < stm32_exti_nbcpus(exti); i++)
+		exti->c_cids[i] = io_read32(exti->base + _EXTI_CmCIDCFGR(i));
+}
+
 static void stm32_exti_pm_suspend(struct stm32_exti_pdata *exti)
 {
 	uint32_t base = exti->base;
 	uint32_t mask = 0;
 	uint32_t i = 0;
 
+	if (IS_ENABLED(CFG_STM32_RIF) && stm32_exti_maxcid(exti))
+		stm32_exti_rif_save(exti);
+
 	for (i = 0; i < _EXTI_BANK_NR; i++) {
 		/* Save ftsr, rtsr and seccfgr registers */
 		exti->ftsr_cache[i] = io_read32(base + _EXTI_FTSR(i));
@@ -249,6 +462,9 @@ static void stm32_exti_pm_resume(struct stm32_exti_pdata *exti)
 	/* Restore EXTI port selection */
 	for (i = 0; i < _EXTI_MAX_CR; i++)
 		io_write32(base + _EXTI_CR(i), exti->port_sel_cache[i]);
+
+	if (IS_ENABLED(CFG_STM32_RIF) && stm32_exti_maxcid(exti))
+		stm32_exti_rif_apply(exti);
 }
 
 static TEE_Result
@@ -300,6 +516,14 @@ static TEE_Result stm32_exti_probe(const void *fdt, int node,
 	exti->base = io_pa_or_va_secure(&base, dt_info.reg_size);
 	assert(exti->base);
 
+	exti->hwcfgr1 = io_read32(exti->base + _EXTI_HWCFGR1);
+	if (IS_ENABLED(CFG_STM32_RIF) && stm32_exti_maxcid(exti)) {
+		stm32_exti_rif_parse_dt(exti, fdt, node);
+		res = stm32_exti_rif_apply(exti);
+		if (res)
+			goto err;
+	}
+
 	res = dt_driver_register_provider(fdt, node,
 					  (get_of_device_func)
 					  stm32_exti_get_handle,
@@ -314,6 +538,8 @@ static TEE_Result stm32_exti_probe(const void *fdt, int node,
 	return TEE_SUCCESS;
 
 err:
+	free(exti->e_cids);
+	free(exti->c_cids);
 	free(exti);
 	return res;
 }
-- 
2.47.2

